core/src/main/java/org/infinispan/util/concurrent/locks/impl/DefaultPendingLockManager.java

```diff
@@ -0,0 +1,373 @@
+      if (transactions.isEmpty()) {
+         return PendingLockPromise.NO_OP;
+      }
+
+      PendingLockPromiseImpl pendingLockPromise = new PendingLockPromiseImpl(transactions, timeService.expectedEndTime(time, unit));
+      PendingLockPromiseImpl existing = pendingLockPromiseMap.putIfAbsent(globalTransaction, pendingLockPromise);
+      if (existing != null) {
+         return existing;
+      }
+      pendingLockPromise.registerListenerInCacheTransactions();
+      if (!pendingLockPromise.isReady()) {
+         timeoutExecutor.schedule(pendingLockPromise, time, unit);
+      }
+      return pendingLockPromise;
+   }
+
+   private int getTopologyId(TxInvocationContext<?> context) {
+      final CacheTransaction tx = context.getCacheTransaction();
+      boolean isFromStateTransfer = context.isOriginLocal() && ((LocalTransaction) tx).isFromStateTransfer();
+      // if the transaction is from state transfer it should not wait for the backup locks of other transactions
+      if (!isFromStateTransfer) {
+         final int transactionTopologyId = tx.getTopologyId();
+         if (transactionTopologyId != TransactionTable.CACHE_STOPPED_TOPOLOGY_ID) {
+            if (transactionTable.getMinTopologyId() < transactionTopologyId) {
+               return transactionTopologyId;
+            }
+         }
+      }
+      return NO_PENDING_CHECK;
+   }
+
+   private long checkForPendingLock(Object key, GlobalTransaction globalTransaction, int transactionTopologyId, long lockTimeout) throws InterruptedException {
+      if (trace) {
+         log.tracef("Checking for pending locks and then locking key %s", toStr(key));
+      }
+
+      final long expectedEndTime = timeService.expectedEndTime(lockTimeout, TimeUnit.MILLISECONDS);
+
+      final CacheTransaction lockOwner = waitForTransactionsToComplete(getTransactionWithLockedKey(transactionTopologyId, key, globalTransaction),
+                                                                       expectedEndTime);
+
+      // Then try to acquire a lock
+      if (trace) {
+         log.tracef("Finished waiting for other potential lockers, trying to acquire the lock on %s", toStr(key));
+      }
+
+      if (lockOwner != null) {
+         timeoutForKey(key, lockOwner, globalTransaction);
+      }
+
+      return timeService.remainingTime(expectedEndTime, TimeUnit.MILLISECONDS);
+   }
+
+   private long checkForPendingAnyLocks(Collection<Object> keys, GlobalTransaction globalTransaction, int transactionTopologyId, long lockTimeout) throws InterruptedException {
+      if (trace)
+         log.tracef("Checking for pending locks and then locking key %s", toStr(keys));
+
+      final long expectedEndTime = timeService.expectedEndTime(lockTimeout, TimeUnit.MILLISECONDS);
+
+      final CacheTransaction lockOwner = waitForTransactionsToComplete(getTransactionWithAnyLockedKey(transactionTopologyId, keys, globalTransaction),
```

Comment:
Thanks!  Makes sense, I am not as familiar with the locking deep down nitty gritty details.